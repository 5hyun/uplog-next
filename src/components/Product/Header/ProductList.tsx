import { AiOutlinePlus } from 'react-icons/ai';
import React, { useCallback, useEffect } from 'react';
import { product, productOpen } from '@recoil/Product/atom.tsx';
import { useRecoilState } from 'recoil';
import { RxDragHandleDots2 } from 'react-icons/rx';
import { BiPencil } from 'react-icons/bi';
import { DragDropContext, Draggable, Droppable, DropResult } from 'react-beautiful-dnd';

export default function ProductList() {
  const [productList, setProductList] = useRecoilState(product);

  const [isProductClick] = useRecoilState(productOpen);

  useEffect(() => {
    if (!isProductClick) return;

    document.body.style.cssText = `
    position: fixed;
    top: -${window.scrollY}px;
    overflow-y: scroll;
    width: 100%;`;
    return () => {
      const scrollY = document.body.style.top;
      document.body.style.cssText = '';
      window.scrollTo(0, parseInt(scrollY || '0', 10) * -1);
    };
  }, [isProductClick]);

  // 제품 list 클릭해도 꺼지지않게함
  const onChildClick = (e: React.MouseEvent) => {
    e.stopPropagation();
  };

  // const onDragEnd = useCallback(
  //   (result: DropResult) => {
  //     console.log(result);
  //     const { destination, draggableId, source } = result;
  //     console.log('drag', draggableId);
  //     console.log('des', destination);
  //
  //     // 이상한 곳에 드래그하면 return
  //     if (!destination) return;
  //
  //     // 출발지와 도착지가 다르면 재정렬
  //     // 깊은 복사
  //     const updatedProduct = JSON.parse(JSON.stringify(product)) as typeof product;
  //     // 기존 아이템 뽑아내기
  //     const [movedItem] = updatedProduct.splice(source.index, 1);
  //     // 기존 아이템을 새로운 위치에 삽입하기
  //     updatedProduct.splice(destination.index, 0, movedItem);
  //
  //     // 상태 변경
  //     console.log('updateList', updatedProduct);
  //     setProductList(updatedProduct);
  //   },
  //   [product]
  // );

  const onDragEnd = useCallback(
    (result: DropResult) => {
      const { destination, source } = result;

      // 이상한 곳에 드래그하면 return
      if (!destination) return;

      // 출발지와 도착지가 다르면 재정렬
      // 깊은 복사
      if (Array.isArray(productList)) {
        const updatedProduct = JSON.parse(JSON.stringify(productList)) as typeof productList;
        // 기존 아이템 뽑아내기
        const [movedItem] = updatedProduct.splice(source.index, 1);
        // 기존 아이템을 새로운 위치에 삽입하기
        updatedProduct.splice(destination.index, 0, movedItem);

        // 상태 변경
        setProductList(updatedProduct);
      }
    },
    [productList]
  );

  return (
    <section
      className={
        'border-base w-[20rem] min-h-[3.3rem] max-h-[27.3rem] block absolute top-[4rem] right-[-8.5rem] shadow-sign-up'
      }
      onClick={onChildClick}
    >
      <DragDropContext onDragEnd={(result) => onDragEnd(result)}>
        <div className={'max-h-[23.7rem] overflow-y-auto'}>
          {/*제품 list*/}
          <Droppable droppableId="droppable">
            {(provided) => (
              <div
                ref={provided.innerRef}
                {...provided.droppableProps}
                className={'flex-col-center justify-between w-full'}
              >
                {productList.map((product, index) => {
                  return (
                    <Draggable
                      draggableId={product.draggableId}
                      key={product.draggableId}
                      index={index}
                    >
                      {(provided) => (
                        <div
                          ref={provided.innerRef}
                          {...provided.draggableProps}
                          {...provided.dragHandleProps}
                          className={
                            'flex-row-center justify-between w-full h-[4.5rem] hover:bg-hover'
                          }
                        >
                          <RxDragHandleDots2
                            className={'flex w-[2.6rem] items-center text-4xl ml-4 fill-gray-light'}
                          />
                          <img
                            src={product.image}
                            alt="userprofile"
                            className={'ml-2 w-[2rem] h-[2rem]'}
                          />
                          <span className={'ml-3 text-xl font-bold w-full'}>{product.name}</span>
                          <BiPencil className={'w-20 text-xl mr-4 fill-gray-light'} />
                        </div>
                      )}
                    </Draggable>
                  );
                })}
                {provided.placeholder}
              </div>
            )}
          </Droppable>
        </div>
      </DragDropContext>

      {/*제품 추가하기*/}
      <div
        className={
          'flex-row-center justify-between w-full h-[3.3rem] border-solid border-t border-gray-border'
        }
      >
        <AiOutlinePlus className={'flex items-center text-2xl ml-4 fill-gray-light'} />
        <span className={'flex items-center w-full text-base ml-3 text-gray-light font-medium'}>
          제품 추가하기
        </span>
      </div>
    </section>
  );
}
