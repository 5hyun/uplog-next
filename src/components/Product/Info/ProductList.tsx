import { AiOutlinePlus } from 'react-icons/ai';
import React, { useCallback, useState } from 'react';
import { productOpen } from '@/recoil/Product/atom.ts';
import { useRecoilState } from 'recoil';
import { RxDragHandleDots2 } from 'react-icons/rx';
import { BiPencil } from 'react-icons/bi';
import { DragDropContext, Draggable, Droppable, DropResult } from 'react-beautiful-dnd';
import { useDisclosure } from '@chakra-ui/react';
import ProductInfoModal from '@/components/Product/Info/ProductInfoModal.tsx';
import { Scrollbars } from 'rc-scrollbars';
import { useNavigate } from 'react-router-dom';
import { SaveUserInfo } from '@/typings/member.ts';
import { useMutation, useQuery, useQueryClient } from 'react-query';
import { changeProductsSequence, getMyProducts } from '@/api/Product/Product.ts';
import { useGetAllProduct } from '@/components/Product/hooks/useGetAllProduct.ts';

export default function ProductList() {
  const navigate = useNavigate();
  const userInfo: SaveUserInfo = JSON.parse(sessionStorage.getItem('userInfo')!);

  // 수정할 product id
  const [productId, setProductId] = useState(1);

  // ProductList 정보
  // const [productList, setProductList] = useRecoilState(productListData);

  // 제품 List click
  const [isProductClick, setIsProductClick] = useRecoilState(productOpen);
  const onCloseProduct = useCallback(() => {
    setIsProductClick(false);
  }, []);

  // 제품추가&정보수정 모달창
  const { isOpen, onOpen, onClose } = useDisclosure();
  const [isCreateProduct, setIsCreateProduct] = useState(true);

  const [productList, refetch] = useGetAllProduct();

  console.log(productList);

  const { mutate } = useMutation(changeProductsSequence);

  // 제품 list 클릭해도 꺼지지않게함
  const onChildClick = (e: React.MouseEvent) => {
    e.stopPropagation();
  };

  // 제품 추가인지 수정인지 체크 여부
  const onChangeIsCreateProduct = useCallback((check: boolean) => {
    setIsCreateProduct(check);
  }, []);

  const queryClient = useQueryClient();

  // dnd - 드래그 끝나면 실행되는 함수
  const onDragEnd = useCallback(
    (result: DropResult) => {
      const { destination, source } = result;

      // 이상한 곳에 드래그하면 return
      if (!destination) return;

      // 출발지와 도착지가 다르면 재정렬
      if (Array.isArray(productList)) {
        // 깊은 복사
        const updatedProduct = JSON.parse(JSON.stringify(productList)) as typeof productList;
        // 기존 아이템 뽑아내기
        const [movedItem] = updatedProduct.splice(source.index, 1);
        // 기존 아이템을 새로운 위치에 삽입하기
        updatedProduct.splice(destination.index, 0, movedItem);

        const updateIndexList: number[] = updatedProduct.map((item) => item.productId);
        mutate(updateIndexList);

        // 바로 쿼리에 업데이트
        queryClient.setQueriesData('myProductList', updatedProduct);
      }
    },
    [productList]
  );

  return (
    <section
      className={
        'border-base w-[20rem] min-h-[3.3rem] max-h-[27.3rem] block absolute top-[4.5rem] left-[11rem] shadow-sign-up z-50'
      }
      onClick={onChildClick}
    >
      <DragDropContext onDragEnd={(result) => onDragEnd(result)}>
        <Scrollbars
          autoHeight
          autoHeightMax={'23.7rem'}
          autoHide
          autoHideTimeout={1000}
          // Duration for hide animation in ms.
          autoHideDuration={200}
        >
          {/*제품 list*/}
          <Droppable droppableId="droppable">
            {(provided) => (
              <div
                ref={provided.innerRef}
                {...provided.droppableProps}
                className={'flex-col-center justify-between w-full'}
              >
                {productList?.map((product, index) => {
                  return (
                    <Draggable
                      draggableId={product.draggableId}
                      key={product.draggableId}
                      index={index}
                    >
                      {(provided, snapshot) => (
                        <div
                          ref={provided.innerRef}
                          {...provided.draggableProps}
                          {...provided.dragHandleProps}
                          className={`flex-row-center justify-between w-full h-[4.5rem] hover:bg-hover ${
                            snapshot.isDragging ? 'shadow-2xl shadow-gray-400' : ''
                          }`}
                        >
                          <div
                            className={'flex-row-center h-full w-[85%]'}
                            onClick={() => {
                              navigate(`/workspace/${product.productId}`);
                              sessionStorage.setItem('nowProduct', JSON.stringify(product));
                              onCloseProduct();
                            }}
                          >
                            <RxDragHandleDots2
                              className={
                                'flex w-[2.6rem] items-center text-4xl ml-4 fill-gray-light'
                              }
                              onClick={() => setIsProductClick(!isProductClick)}
                            />
                            <img
                              src={product.image}
                              alt="userprofile"
                              className={'ml-2 w-[2rem] h-[2rem]'}
                              onClick={() => setIsProductClick(!isProductClick)}
                            />
                            <span
                              className={'ml-3 text-xl font-bold w-full'}
                              onClick={() => setIsProductClick(!isProductClick)}
                            >
                              {product.productName}
                            </span>
                          </div>

                          {(product.powerType === 'MASTER' || product.powerType === 'LEADER') && (
                            <span
                              className={'flex-row-center w-[15%] h-full'}
                              onClick={() => {
                                onOpen();
                                onChangeIsCreateProduct(false);
                                setProductId(product.productId);
                              }}
                            >
                              <BiPencil
                                className={'text-xl fill-gray-light cursor-pointer z-50 '}
                              />
                            </span>
                          )}
                        </div>
                      )}
                    </Draggable>
                  );
                })}
                {provided.placeholder}
              </div>
            )}
          </Droppable>
        </Scrollbars>
      </DragDropContext>

      {/*제품 추가하기, 기업만 생성버튼 보임*/}
      {userInfo.position === 'COMPANY' && (
        <div
          className={
            'flex-row-center justify-between w-full h-[3.3rem] border-solid border-t border-gray-border'
          }
          onClick={() => {
            onOpen();
            onChangeIsCreateProduct(true);
          }}
        >
          <AiOutlinePlus className={'flex items-center text-2xl ml-4 fill-gray-light'} />
          <span className={'flex items-center w-full text-base ml-3 text-gray-light font-medium'}>
            제품 추가하기
          </span>
        </div>
      )}
      <ProductInfoModal
        isOpen={isOpen}
        onClose={onClose}
        isCreateProduct={isCreateProduct}
        productId={productId}
      />
    </section>
  );
}
