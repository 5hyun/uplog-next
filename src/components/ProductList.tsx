import { AiOutlinePlus } from 'react-icons/ai';
import React, { useCallback, useEffect, useRef, useState } from 'react';
import { productOpen } from '../recoil/Product/atom.tsx';
import { useRecoilState } from 'recoil';
import { RxDragHandleDots2 } from 'react-icons/rx';
import { BiPencil } from 'react-icons/bi';
import { DragDropContext, Draggable, Droppable, DropResult } from 'react-beautiful-dnd';
import { ProductInfo } from '../types/product.ts';

export default function ProductList() {
  const product: ProductInfo[] = [
    {
      id: 1,
      draggableId: '1',
      name: 'Product1',
      image: '/images/test_userprofile.png',
      master: 'master',
      client: 'client',
    },
    {
      id: 2,
      draggableId: '2',
      name: 'Product2',
      image: '/images/test_userprofile.png',
      master: 'master',
      client: 'client',
    },
    {
      id: 3,
      draggableId: '3',
      name: 'Product3',
      image: '/images/test_userprofile.png',
      master: 'master',
      client: 'client',
    },
    {
      id: 4,
      draggableId: '4',
      name: 'Product4',
      image: '/images/test_userprofile.png',
      master: 'master',
      client: 'client',
    },
    {
      id: 5,
      draggableId: '5',
      name: 'Product5',
      image: '/images/test_userprofile.png',
      master: 'master',
      client: 'client',
    },
  ];
  const getItemStyle = (isDragging: any, draggableStyle: any) => ({
    ...draggableStyle,
  });

  const [productList, setProductList] = useState<ProductInfo[]>(product);

  const [isProductClick, setIsProductClick] = useRecoilState(productOpen);

  useEffect(() => {
    if (!isProductClick) return;

    document.body.style.cssText = `
    position: fixed;
    top: -${window.scrollY}px;
    overflow-y: scroll;
    width: 100%;`;
    return () => {
      const scrollY = document.body.style.top;
      document.body.style.cssText = '';
      window.scrollTo(0, parseInt(scrollY || '0', 10) * -1);
    };
  }, [isProductClick]);

  // 제품 list 클릭해도 꺼지지않게함
  const onChildClick = (e: React.MouseEvent) => {
    e.stopPropagation();
  };

  const onDragEnd = useCallback(
    (result: DropResult) => {
      console.log(result);
      const { destination, draggableId, source } = result;
      console.log('des', destination);
      console.log('drag', draggableId);

      // 이상한 곳에 드래그하면 return
      if (!destination) return;

      // 출발지와 도착지가 같으면 return
      if (destination.droppableId === source.droppableId && source.index === destination.index)
        return;

      // 출발지와 도착지가 다르면 재정렬
      const updatedProduct = Array.from(product);
      const [movedItem] = updatedProduct.splice(source.index, 1);
      updatedProduct.splice(destination.index, 0, movedItem);

      // 이제 'updatedProduct'에는 항목이 새 인덱스로 이동한 배열이 들어 있습니다.
      console.log('재정렬된 물품:', updatedProduct);
      setProductList(updatedProduct);
    },
    [product]
  );

  return (
    <section
      className={
        'border-base w-[20rem] min-h-[3.3rem] max-h-[30rem] block absolute top-[4rem] right-[-8.5rem] shadow-sign-up'
      }
      onClick={onChildClick}
    >
      <div className={'max-h-[26.7rem] overflow-y-auto'}>
        {/*제품 list*/}
        <DragDropContext onDragEnd={(result) => onDragEnd(result)}>
          <Droppable droppableId="droppable" direction="vertical">
            {(provided) => (
              <div
                ref={provided.innerRef}
                {...provided.droppableProps}
                className={'flex-col-center justify-between w-full h-[4.5ㅎrem] hover:bg-hover'}
              >
                {productList.map((product, index) => {
                  return (
                    <Draggable draggableId={product.draggableId} index={index}>
                      {(provided, snapshot) => (
                        <div
                          ref={provided.innerRef}
                          {...provided.draggableProps}
                          {...provided.dragHandleProps}
                          // style={getItemStyle(snapshot.isDragging, provided.draggableProps.style)}
                          className={'flex-row-center justify-between w-full h-[4.5rem]'}
                        >
                          <RxDragHandleDots2
                            className={'flex w-[2.6rem] items-center text-4xl ml-4 fill-gray-light'}
                          />
                          <img
                            src={product.image}
                            alt="userprofile"
                            className={'ml-2 w-[2rem] h-[2rem]'}
                          />
                          <span className={'ml-3 text-xl font-bold w-full'}>{product.name}</span>
                          <BiPencil className={'w-20 text-xl mr-4 fill-gray-light'} />
                        </div>
                      )}
                    </Draggable>
                  );
                })}
                {provided.placeholder}
              </div>
            )}
          </Droppable>
        </DragDropContext>
      </div>

      {/*제품 추가하기*/}
      <div
        className={
          'flex-row-center justify-between w-full h-[3.3rem] border-solid border-t border-gray-border'
        }
      >
        <AiOutlinePlus className={'flex items-center text-2xl ml-4 fill-gray-light'} />
        <span className={'flex items-center w-full text-base ml-3 text-gray-light font-medium'}>
          제품 추가하기
        </span>
      </div>
    </section>
  );
}
