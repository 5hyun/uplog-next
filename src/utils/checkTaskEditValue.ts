import { TaskData } from '@/typings/task.ts';

export function checkTaskEditValue(editTask: TaskData): boolean {
  if (editTask.taskName === '') {
    return false;
  }
  if (editTask.targetMemberInfoDTO && editTask.targetMemberInfoDTO.id === 0) {
    return false;
  }
  if (editTask.menuId === 0) {
    return false;
  }
  if (editTask.teamId === 0) {
    return false;
  }
  if (editTask.startTime === '' || editTask.endTime === '') {
    return false;
  }

  return true;
}
