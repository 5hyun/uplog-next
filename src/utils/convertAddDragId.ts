import { StatusTaskData, Tasks } from '@/typings/task.ts';
import { useState } from 'react';

export function convertAddDragId(tasks: StatusTaskData): Tasks {
  const [convertedTasks, setConvertedTasks] = useState<Tasks>({
    PROGRESS_BEFORE: [],
    PROGRESS_IN: [],
    PROGRESS_COMPLETE: [],
  });

  const beforeStatusTasks = tasks.PROGRESS_BEFORE.map((task) => ({
    ...task,
    dragId: task.id.toString(), // dragId 값을 문자열로 변환하여 할당
  }));

  setConvertedTasks({
    ...convertedTasks,
    PROGRESS_BEFORE: beforeStatusTasks,
  });

  const inStatusTasks = tasks.PROGRESS_IN.map((task) => ({
    ...task,
    dragId: task.id.toString(), // dragId 값을 문자열로 변환하여 할당
  }));

  setConvertedTasks({
    ...convertedTasks,
    PROGRESS_IN: inStatusTasks,
  });

  const completeStatusTasks = tasks.PROGRESS_COMPLETE.map((task) => ({
    ...task,
    dragId: task.id.toString(), // dragId 값을 문자열로 변환하여 할당
  }));

  setConvertedTasks({
    ...convertedTasks,
    PROGRESS_COMPLETE: completeStatusTasks,
  });

  return convertedTasks;
}
