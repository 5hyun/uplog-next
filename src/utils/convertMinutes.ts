// ** 타이머 분:초 구하는 함수 */
export function convertMinutes(seconds: number): string {
  const minutes = Math.floor(seconds / 60);
  const remainingSeconds = seconds % 60;
  const formattedSeconds = remainingSeconds < 10 ? `0${remainingSeconds}` : `${remainingSeconds}`;

  if (minutes === 0) {
    return `00:${formattedSeconds}`;
  }

  return `${minutes}:${formattedSeconds}`;
}
