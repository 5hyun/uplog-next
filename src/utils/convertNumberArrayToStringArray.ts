export function convertNumberArrayToStringArray(numbers: number[]): string[] {
  // 배열의 각 요소를 문자열로 변환하여 새로운 배열을 생성합니다.
  const strings: string[] = numbers.map((number) => number.toString());
  return strings;
}
