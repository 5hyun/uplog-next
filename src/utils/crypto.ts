import CryptoJS from 'crypto-js';

/* Salt 비밀 키 */
// const salt = import.meta.env.VITE_SALT;
const salt = '1234';

/* 암호화 */
export const encrypt = (text: string) => {
  // 값이 없을 경우 빈 문자열 반환
  if (!text) return '';
  return CryptoJS.AES.encrypt(text, salt).toString();
};

/* 복호화 */
export const decrypt = (text: string) => {
  // 값이 없을 경우 빈 문자열 반환
  if (!text) return '';

  try {
    const bytes = CryptoJS.AES.decrypt(text, salt); // 복호화 시도
    return bytes.toString(CryptoJS.enc.Utf8);
  } catch (error) {
    return ''; // 에러 발생 시 빈 문자열 반환
  }
};
