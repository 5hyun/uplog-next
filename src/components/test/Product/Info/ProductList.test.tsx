import { describe, expect, test } from '@jest/globals';
import { fireEvent, getByText, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { withRouter } from '@/tests/utils';
import { Route, Router } from 'react-router-dom';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import ProductList from '@/components/Product/Info/ProductList';

describe('ProductList', () => {
  test('ProductList UI', () => {
    const component = renderer.create(
      withRouter(<Route path={'/'} element={<ProductList />} />, '/workspace/product')
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
