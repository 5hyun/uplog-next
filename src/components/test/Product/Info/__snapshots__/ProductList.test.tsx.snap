// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList ProductList UI 1`] = `null`;
